Switch v2 libraries to Learning Core data models

ormsbee opened the pull request:
## Description & Supporting Information

This moves the Content Libraries V2 backend from [Blockstore](https://github.com/openedx-unsupported/blockstore) over to [Learning Core](https://github.com/openedx/openedx-learning/). For high-level overview and rationale of this move, see

- https://github.com/openedx/public-engineering/issues/238

Supersedes: 

- https://github.com/openedx/edx-platform/pull/33577 and
- https://github.com/openedx/edx-platform/pull/33579.

Requires:

- [x] https://github.com/openedx/openedx-learning/pull/143
- [x] https://github.com/openedx/openedx-learning/pull/148
- [x] https://github.com/openedx/openedx-learning/pull/149

### A note on backwards incompatibility

On the Blockstore DEPR, we did not hear that anyone is relying on Blockstore-backed V2 Content Libraries in production. So, we are applying this port in-place without any migration path from Blockstore to Learning Core. **If you were using V2 Content Libraries in production: your libraries will stop working.** However, we are not yet dropping the the `ContentLibraries.bundle_uuid` field, so it is still possible to manually migrate Blockstore data over to Learning Core. If you're in the situation, reach out to us. The `ContentLibraries.bundle_uuid` field will be dropped in Sumac.

## Status

The biggest functional gap here is static asset support. I'm gong to try to merge this once I have tests fixed, even if it means we take a hit in terms of a regression in some functionality like static asset support–with the understanding that we're going to keep iterating to make this ready for Redwood.

TODO in this PR:

- [x] Convert remaining Learning Core model manipulation to use `api` module calls instead.
- [x] Squash the new migrations into one.
- [x] So much test removal/cleanup/rewriting.

TODO in follow-up PRs:

* https://github.com/openedx/edx-platform/issues/34283

Bugs I'm seeing (not sure if I introduced them or not):

* There's sometimes a JS error when a new block is created, about not being able to set `display_name` on null. But the value is written correctly into the database and shows up correctly when reloading.
* The ProblemBlock editor seems to revert to code editing after the initial creation–did I change the OLX serialization? 

FYI @kdmccormick, @bradenmacdonald 

## Testing

This PR's sandbox can be compared against:

* https://github.com/openedx/edx-platform/pull/33936

If you want to run this on your devstack, you must put the following in your `cms/envs/private.py`:

```python
from .devstack import FEATURES

FEATURES['ENABLE_LIBRARY_AUTHORING_MICROFRONTEND'] = True
```

### PR sandbox

**Settings**
```yaml
PLUGINS:
    - mfe
    - grove
    - s3
    - library-authoring-mfe
```
**Tutor requirements**
```
git+https://github.com/overhangio/tutor.git@nightly
git+https://github.com/overhangio/tutor-mfe.git@nightly
git+https://gitlab.com/opencraft/dev/tutor-contrib-grove.git@main
git+https://github.com/hastexo/tutor-contrib-s3.git@main
git+https://github.com/openedx/openedx-tutor-plugins@main#subdirectory=plugins/tutor-contrib-library-authoring-mfe
```

ormsbee committed:
feat: convert v2 content libraries to use Learning Core models

ormsbee commented:
Okay, I've updated to the changes in https://github.com/openedx/openedx-learning/pull/149, and it seems to be working locally. In the morning, I'll squash what I have here, do a rebase to what's on master today, and then see what explodes in terms of conflicts.

ormsbee committed:
refactor: no need to spell out the app config when there's only one

ormsbee committed:
chore: remove comment about libraries being blockstore-based

ormsbee committed:
test: fix model tests for content_libraries

ormsbee committed:
test: remove unnecessary mixin layering of test; remove parent/child tests

(we don't support that)

ormsbee committed:
fix: fix the filtering test case (requires change from openedx-learning)

ormsbee committed:
test: remove links-related tests (we no longer have them)

ormsbee committed:
test: remove tests regarding library-type conversion (we're no longer supporting that).

ormsbee committed:
test: more text patching

ormsbee committed:
fix: fix default xblock title behavior and adjust more tests

ormsbee committed:
fix: deleted blocks should 404 instead of 500

ormsbee committed:
feat!: remove blockstore lib import to learning core

There are no live v2 blockstore libraries that we're planning to
migrate.

ormsbee committed:
refactor: adjust openedx-learning usage to match 0.6.0 changes

ormsbee committed:
test: fix content_tagging tests to remove blockstore collection use

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -294,58 +281,40 @@ def get_libraries_for_user(user, org=None, library_type=None):
-    ]
 
     libraries = [
+        # TODO: Do we really need these fields for the library listing view?
```
Personally, I would say that this is definitely "nice to have" in the UI  (indicate which libraries have unfinished changes, so you can find them and finish your work), but definitely not a requirement for MVP.

But there are other ways to work around this from a UX perspective - for example, trying to go for something more like Google Docs, where changes are encouraged to be published immediately and if you close a library page / XBlock editor without publishing your changes to any given XBlock it will prompt you "are you sure you want to leave without publishing your changes?". In that scenario, unpublished changes are rare and we don't need to highlight them as much.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -373,20 +342,43 @@ def get_library(library_key):
+                              ).exists()
+
+    # I'm doing this one to match what was already existing, but I'm not clear
+    # on the use case for it.
```
Here are the assumptions that went into this:

1. Eventually, the published version of all libraries should match the draft version. If they are different in any way, that indicates that the library is "in development", and some author needs to give it attention and finish the work they were doing.
2. Thus, if there is unfinished work in a library (it's "in development"), we need to indicate that in the UI and draw it to the attention of the authors so they can finish that work. (This would happen on the library list view)
3. Due to how blockstore was implemented, we cannot just check "has changes" (draft != published) for a library, but we [have to separately check both for unpublished changes to remaining files and unpublished deletes](https://github.com/openedx/edx-platform/blob/5023e6901d87327fcea4bf0980d047d38603d762/openedx/core/djangoapps/content_libraries/library_bundle.py#L282-L329). AFAIK nobody actually cares about distinguishing those two things; on the list view, we just want to know for the entire library if draft != published. And on the "library components" view, it would ideally indicate which components have unpublished changes or have been deleted but not published. (I don't think that has been implemented yet.)


Some of these assumptions may no longer hold. Related, I think that having a "Publish" button for the library as a whole is dumb, and there should be instead per-Component "Publish" buttons as well as perhaps a "Publish All". Or as I said, it may be an even better approach to heavily encourage the user to publish changes as quickly as possible, especially if there are no good tools for viewing diffs, historical versions, etc.

kdmccormick replied:
Dave noted this in a TODO comment

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -373,20 +342,43 @@ def get_library(library_key):
+    # we've decided to do those version references at the level of the
+    # individual blocks being used, since a Learning Core backed library is
+    # intended to be used for many LibraryContentBlocks and not 1:1 like v1
+    # libraries
```
👍🏻 I think it's sufficient to return only the last published date for now; I don't think we need a version. We can revisit it if/when we get into cross-instance content sharing.

ormsbee replied:
Right now, we need it because the `LibraryContentBlock` wants something to put for the library's version, and I figure that we might as well pass it through the REST API as well. But yeah, another one for the longer term chopping block.

kdmccormick replied:
I added this to our TODO list. If the library-wide version is on the chopping block, I want to get this out of the REST API before anyone starts relying on it :)

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -553,89 +542,67 @@ def update_library(
+            if allow_public_read is not None:
+                content_lib.allow_public_read = allow_public_read
+            if library_type is not None:
+                # TODO: Can we get rid of this field entirely?
```
> `# TODO: Can we get rid of this field entirely?`

Yes. Feel free to do it now.

ormsbee replied:
I'm punting on removing it for this PR, but it's on my chopping list.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -647,98 +614,81 @@ def _get_library_component_tags_count(library_key) -> dict:
-
 def set_library_block_olx(usage_key, new_olx_str):
     """
+    TODO: When is it even being called now?
```
Someone removed [the "OLX" tab](https://github.com/open-craft/ramshackle/blob/14e0b86029793f5901e3c4edb0e7891d53ef4c56/docs/screenshot4.png) that I had put into the new Library Authoring Frontend, but I do hope that it comes back at some point. It allows advanced users to just edit the OLX directly in the UI, including functionality like preserving comments in the OLX.

Additionally, there is a REST API for this method, and I think some of the new editors were using it at one point.

ormsbee replied:
A number of unit tests used this to set up their test data as well, it turns out. Which makes perfect sense.

bradenmacdonald commented on openedx/core/djangoapps/xblock/api.py:
```diff
@@ -50,6 +55,10 @@ def get_runtime_system():
     """
+    # TODO: Is any of the following necessary now that we're no longer using
+    # Blockstore or its caching mechanisms? And why were we doing a dict with
+    # attributes manually set by thread ID instead of a ContextVar?
```
If we're not supporting child blocks in content libraries, this is probably not necessary anymore. But if you are supporting child blocks, and you're in any way caching the resulting [`.children`] field data keyed by XBlock usage or definition key, it becomes difficult to avoid. With Learning Core though, maybe we don't need such a field data cache. If so, this isn't needed.

ormsbee committed:
test: remove test for v2 lib child block functionality (now unsupported)

ormsbee committed:
test: remove blockstore usage from clipboard paste test

ormsbee commented:
There's one major outstanding bug that I'm seeing, and that is that it looks like the Learning Core XBlock runtime is serializing the default values into the metadata dict that comes across the fields REST API call that the editor uses.

Looking back, the XBlock runtime for Learning Core is something I copy-pasta'd from Blockstore, with some things slashed out. There are parts of the original blockstore runtime that I don't really understand as well as I should. @bradenmacdonald: May I grab an hour of your time on Monday to walk through the Blockstore XBlock runtime and FieldData?

ormsbee commented on openedx/core/djangoapps/content_libraries/models.py:
```diff
@@ -97,9 +99,19 @@ class ContentLibrary(models.Model):
     # e.g. "lib:org:slug" is the opaque key for a library.
     org = models.ForeignKey(Organization, on_delete=models.PROTECT, null=False)
     slug = models.SlugField(allow_unicode=True)
-    bundle_uuid = models.UUIDField(unique=True, null=False)
```
@kdmccormick, @bradenmacdonald: I took this out mostly so I could find code and tests that accessed it. But maybe I should leave `bundle_uuid` in until after Redwood? Just to preserve the blockstore data relationship in case someone did somehow use it?

bradenmacdonald replied:
Yeah, if you want to be extra careful, you could make it nullable and rename the python version of the column, without renaming the DB column, to get that same benefit of removing all code usage without removing the DB data just yet.

ormsbee committed:
fix: remove atomic decorator from views

ormsbee committed:
refactor: shift library search logic from view to api

ormsbee committed:
fix: force content libraries views to be nonatomic

ormsbee committed:
refactor: remove unnecessary imports

ormsbee commented:
@kdmccormick, @bradenmacdonald: The `openedx/core/djangoapps/content_libraries/views.py` module is ready for review.

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at https://pr-34066-2432b1.sandboxes.opencraft.hosting
Sandbox Studio is available at https://studio.pr-34066-2432b1.sandboxes.opencraft.hosting

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6156678534-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6156678534-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6156678534-tutor-requirements.txt

ormsbee committed:
fix: have field data properly track changed fields (and simplify the runtime/fielddata)

ormsbee committed:
fix: remove error log that I was only using for debugging purposes

ormsbee committed:
refactor: cleanups, comments

ormsbee committed:
refactor: more comment tweaks

ormsbee commented:
@kdmccormick, @bradenmacdonald: The re-worked `openedx/core/djangoapps/xblock/runtime/learning_core_runtime.py` is ready for review.

ormsbee committed:
refactor: more cleanup of config and removing debug comments

ormsbee committed:
refactor: more cleanup

ormsbee committed:
refactor: remove unused imports

ormsbee committed:
refactor: update comments

kdmccormick commented on openedx/core/djangoapps/xblock/runtime/learning_core_runtime.py:
```diff
@@ -0,0 +1,299 @@
+       blocks have changes to any of their fields. See the marked_unchanged
+       method docstring for more details.
+    """
+    NOT_FOUND = Sentinel('NOT_FOUND')
```
Given that we're raising KeyErrors when something isn't found, is a `NOT_FOUND` sentinel needed? I see it it is used in `set()`, but it looks like that could be factored away easily.

ormsbee replied:
It's a leftover, and I'll get rid of it. Though I'm irrationally fond of the pattern. 😛

ormsbee commented:
Removed the create-sandbox label so that the sandbox gets destroyed. Will remake it with a more backwards compatible migration later.

ormsbee committed:
refactor: remove unnecessary atomic block

kdmccormick commented on openedx/core/djangoapps/xblock/runtime/learning_core_runtime.py:
```diff
@@ -0,0 +1,299 @@
+        # Verify that the user has permission to write to authored data in this
+        # learning context:
+        if self.user is not None:
+            learning_context = get_learning_context_impl(block.scope_ids.usage_id)
```
@ormsbee This returns a subclass of LearningContext, not an instance. Perhaps you just want `learning_context = block.scope_ids.usage_id.context_key`?

ormsbee replied:
I lifted this straight out of the blockstore implementation:

https://github.com/openedx/edx-platform/blob/5a36fa9163dd2608d4089a45ff64c0dec6bc0c9a/openedx/core/djangoapps/xblock/runtime/blockstore_runtime.py#L131-L135

But `block.scope_ids.usage_id.context_key` is the `LearningContextKey` right? Where I need to ultimately get a [`LibraryContextImpl`](https://github.com/openedx/edx-platform/blob/5a36fa9163dd2608d4089a45ff64c0dec6bc0c9a/openedx/core/djangoapps/content_libraries/library_context.py#L21)?

kdmccormick replied:
Right, yup. Also, `get_learning_context_key_impl` does return an instance, despite its docstring. I'm wrong on all counts, please disregard this thread 🙂

kdmccormick commented on openedx/core/djangoapps/xblock/runtime/learning_core_runtime.py:
```diff
@@ -0,0 +1,299 @@
+
+        return component
+
+    def _lookup_asset_url(self, block: XBlock, asset_path: str):  # pylint: disable=unused-argument
```
```suggestion
    def _lookup_asset_url(self, block: XBlock, asset_path: str):  # pylint: disable=unused-argument
```
```suggestion
    def _lookup_asset_url(self, block: XBlock, asset_path: str) -> str | None:  # pylint: disable=unused-argument
```

kdmccormick reviewed (changes requested):
I just reviewed learning_core_runtime.py. Two suggestions, and one thing I think is actually broken (the get_learning_context_impl all).

ormsbee committed:
refactor: add comments to api calls, remove unused function

ormsbee commented:
@kdmccormick, @bradenmacdonald: `openedx/core/djangoapps/content_libraries/api.py` is ready for review.

ormsbee committed:
refactor: changes in response to Kyle's review

ormsbee commented:
@kdmccormick: incorporated your feedback on `learning_core_runtime.py`

ormsbee committed:
refactor: consolidate get_component_from_usage_key implementations

ormsbee committed:
refactor: comment

ormsbee committed:
refactor: remove get_block_display_name

It was previously being called with XBlocks or usage keys, at which
point, it would do a lookup of the display name from the Component. But
now it's only being sent XBlocks, and it only grabs their display_name
when that's the case, so there's no point in keeping it around.

ormsbee commented:
`xblock/api.py` is ready for review.

ormsbee committed:
refactor: remove unnecessary config value for OPENEDX_LEARNING

ormsbee committed:
refactor: remove ENABLE_LIBRARY_AUTHORING_MICROFRONTEND from devstack settings

(you have to add it to private.py)

kdmccormick commented on openedx/core/djangoapps/xblock/api.py:
```diff
@@ -43,31 +50,33 @@
-    The Runtime System isn't always needed (e.g. for management commands), so to
-    keep application startup faster, it's only initialized when first accessed
-    via this method.
+    Return a new XBlockRuntimeSystem.
+
+    TODO: Refactor to get rid of the XBlockRuntimeSystem entirely and just
+    create the LearningCoreXBlockRuntime and return it. We used to want to keep
+    around a long lived runtime system (a factory that returns runtimes) for
+    caching purposes, and have it dynamically construct a runtime on request.
+    Now we're just re-constructing both the system and the runtime in this call
+    and returning it every time, because:
+
+    1. We no longer have slow, Blockstore-style definitions to cache, so the
+       performance of this is perfectly acceptable.
+    2. Having a singleton increases complexity and the chance of bugs.
+    3. Creating the XBlockRuntimeSystem every time only takes about 10-30 µs.
+
+    Given that, the extra XBlockRuntimeSystem class just adds confusion. But
+    despite that, it's tested, working code, and so I'm putting off refactoring
+    for now.
```
Nice writeup. Agreed all around.

kdmccormick commented on openedx/core/djangoapps/xblock/api.py:
```diff
@@ -146,65 +155,57 @@ def get_block_metadata(block, includes=()):
+    )
+
+
+def get_library_block_olx(usage_key: LibraryUsageLocatorV2):
```
```suggestion
def get_library_block_olx(usage_key: LibraryUsageLocatorV2) -> str:
```

ormsbee committed:
fix: make schema migration more backwards-compatible

ormsbee committed:
fix: add LC models to keyword overrides (I'm checking now if this is okay)

ormsbee commented on db_keyword_overrides.yml:
```diff
@@ -16,11 +17,13 @@ MYSQL:
     - OAuth2ProviderConfig.key
     - ProctoredExamStudentAllowanceHistory.key
     - ProctoredExamStudentAllowance.key
+    - PublishableEntity.key
```
Need to verify if this is okay to do, and what the downstream impact is.

ormsbee replied:
We renamed the database columns to `_key` instead. See https://github.com/openedx/openedx-learning/pull/160

ormsbee committed:
fix: address linter errors

ormsbee committed:
fix: more linter fixes

ormsbee committed:
fix: more linter fixes

ormsbee committed:
fix: all the linter things

ormsbee committed:
fix: spacing

ormsbee commented:
force-pushed a rebase onto the latest master

ormsbee committed:
fix: more linter fixes for indentation/spacing

ormsbee committed:
fix: all the sytax linter stuff, I think...

ormsbee committed:
fix: remove entries from db_keyword_overrides (no longer necessary)

ormsbee commented:
@kdmccormick, @bradenmacdonald: Everything here should be reviewable now.

ormsbee committed:
chore: mypy thinks it's a BaseManager and not a ContentLibraryManager

ormsbee commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -649,114 +605,116 @@ def _get_library_component_tags_count(library_key) -> dict:
-        raise ContentLibraryBlockNotFound(usage_key)
-    lib_bundle = LibraryBundle(usage_key.lib_key, def_key.bundle_uuid, draft_name=DRAFT_NAME)
-    return def_key, lib_bundle
+    lib = ContentLibrary.objects.get_by_key(library_key)  # type: ignore[attr-defined]
```
@kdmccormick: This is me giving up. I belatedly realized that the original call before my conversion also had this "type: ignore" on it.

kdmccormick replied:
Np. Looks like django-stubs is pinned to a way old version, which might be the culprit. Now that we're on django 4.2 the pin can be lifted. I'll handle that and then come back and see if this ignore can be removed.

kdmccormick replied:
Will be handled in https://github.com/openedx/edx-platform/pull/34237

ormsbee commented:
The last actual issue I'm aware of is a set of unit tests that fails in CI but not locally. It involves signals and mocks, so I don't know if there's something weird going on there. These same tests were working earlier.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -85,59 +85,26 @@
 )
+from openedx_learning.core.publishing import api as publishing_api
+from openedx_learning.core.contents import api as contents_api
+from openedx_learning.core.contents.models import MediaType
```
Nit: if you just need this for a type annotation, maybe write the type as `content_api.MediaType` and remove this import? Then you're only importing from `api`, which helps enforce boundaries.

Ideally I'd even like to add these learning core apps to [`isolated_apps`](https://github.com/openedx/edx-platform/blob/008467c6e51407187878d8aa1a5453da1af8f591/setup.cfg#L173-L187) so the linter enforces this contract.

kdmccormick replied:
I'll make this change, and I agree on putting boundaries between APIs and other apps' models, but I'm not clear on how making this change will help. We're still passing around MediaType model instances.

bradenmacdonald replied:
Yeah, it's more to be in the habit of only importing from `api.py`, and if we are using a model directly, it should be one exposed via `api.py`. Ideally in the future we could refactor it to use an [OEP-49 data class](https://open-edx-proposals.readthedocs.io/en/latest/best-practices/oep-0049-django-app-patterns.html#data-py) instead of a model.

kdmccormick replied:
Makes sense, changed. If we want to switch Learning Core to exposing data classes instead of models, we should probably do that sooner than later.

ormsbee committed:
fix: skip the libraries test cases under LMS (signals won't get sent)

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -294,58 +255,52 @@ def get_libraries_for_user(user, org=None, library_type=None):
             type=lib.type,
-            description=metadata[i]['description'],
-            version=metadata[i]['version'],
+            description="",
```
Q: For my understanding, why is description omitted if it's already likely loaded as `lib.learning_package.description` ? (Because you want to exclude it from select_related loading in the future perhaps?)

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -649,114 +605,116 @@ def _get_library_component_tags_count(library_key) -> dict:
 
+    TODO: This is primarily used in tests, but it's getting an unbounded list,
+    instead of a pagable queryset. We should get rid of this function
+    altogether.
```
Can't we just merge `get_library_components` into this, and have one function that returns a QuerySet of `LibraryXBlockMetadata` objects? Or is the `Component` model itself necessary for API consumers here?

Edit: (from talking to Dave): it would be nice to just have a QuerySet that returns LibraryXBlockMetadata, but since that's not a django model there's no simple way to achieve that. Definitely something along those lines would be nice refactor in the future. For now, you need to use `get_library_components` and map the results using `LibraryXBlockMetadata.from_component`

kdmccormick replied:
I deleted this function and inlined its definition to the 1 place it's used (the sync-blocks-from-source-library task). I also added a TODO to the PR description.

bradenmacdonald commented on openedx/core/djangoapps/xblock/api.py:
```diff
@@ -146,65 +155,57 @@ def get_block_metadata(block, includes=()):
+    )
+
+
+def get_library_block_olx(usage_key: LibraryUsageLocatorV2):
```
If we're moving this API function to the main `xblock` app, why is it still specific to libraries?

kdmccormick replied:
https://github.com/openedx/edx-platform/pull/34066#discussion_r1491588819

ormsbee commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -649,114 +605,116 @@ def _get_library_component_tags_count(library_key) -> dict:
 def get_library_block(usage_key) -> LibraryXBlockMetadata:
     """
     Get metadata (LibraryXBlockMetadata) about one specific XBlock in a library
 
```
"Get the metadata for the draft version of a block."

ormsbee commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -649,114 +605,116 @@ def _get_library_component_tags_count(library_key) -> dict:
+    new_title = node.attrib.get(
+        "display_name",
+        xblock_type_display_name(usage_key.block_type),
+    )
```
^ Should check that the node tag matches the block type.

kdmccormick replied:
Done. Now raises a ValueError if it doesn't match.

ormsbee commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -781,149 +741,143 @@ def create_library_block(library_key, block_type, definition_id):
+
     # Make sure the proposed ID will be valid:
-    validate_unicode_slug(definition_id)
+    validate_unicode_slug(block_id)
```
@kdmccormick: reminder to check the implementation whether we'll have valid usage keys from this.

kdmccormick replied:
@bradenmacdonald In addition to the usual suspects that `validate_slug/SlugField` checks for (letters, numbers, underscores, hyphens), `validate_unicode_slug` would allow for [modified letters](https://www.compart.com/en/unicode/category/Lm) and [other letters](https://www.compart.com/en/unicode/category/Lo). Even if OpaqueKeys tolerates modified letters and other letters, I'm not sure what all the downstream implications would be.

My sense is to limit it to classic ASCII SlugField for now, and loosen the restriction later if we feel confident. What do you think?

bradenmacdonald replied:
Yes, OpaqueKeys and the content libraries API as it existed before this PR were designed to allow unicode slugs : https://github.com/openedx/opaque-keys/blob/f14bfd7b8970958e0d7f008a73902cb1f8f6686f/opaque_keys/edx/locator.py#L1542-L1543

My preference would be to just change one of the existing full tests like [this](https://github.com/openedx/edx-platform/blob/5ca009ca1bd61e4c433d6d8e51672f1c1ea6c005/openedx/core/djangoapps/content_libraries/tests/test_content_libraries.py#L211-L298) to use a unicode slug, which would basically ensure it works across the whole content libraries lifecycle (except probably wouldn't cover using it in a course). In just the same way as that test currently checks that unicode OLX works.

But I'm also fine with your suggested approach. If you want, you can make it ascii now and add this to the TODO list and I'll work on it in a separate PR with more tests.

kdmccormick replied:
Cool, I was underestimating our ability to handle non-ascii here. I updated that exact test.

ormsbee commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -781,149 +741,143 @@ def create_library_block(library_key, block_type, definition_id):
 
 
-def create_library_block_child(parent_usage_key, block_type, definition_id) -> LibraryXBlockMetadata:
+def create_component_for_block(content_lib, usage_key):
```
Is this being called in a transaction.atomic?

kdmccormick replied:
@bradenmacdonald I don't remember what the discussion here was, do you?

bradenmacdonald replied:
No, sorry.

kdmccormick replied:
I'll add a TODO for Dave. I doubt it needs to block this merge.

kdmccormick replied:
Wrapped this in `with transaction.atomic()`

ormsbee commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -781,149 +741,143 @@ def create_library_block(library_key, block_type, definition_id):
 
 
-def create_library_block_child(parent_usage_key, block_type, definition_id) -> LibraryXBlockMetadata:
+def create_component_for_block(content_lib, usage_key):
```
This is an internal function, not public API.

ormsbee commented on openedx/core/djangoapps/xblock/runtime/id_managers.py:
```diff
@@ -24,7 +21,8 @@ def get_definition_id(self, usage_id):
         if isinstance(usage_id, UsageKeyV2):
-            return get_learning_context_impl(usage_id).definition_for_usage(usage_id)
+            return None
+            # return get_learning_context_impl(usage_id).definition_for_usage(usage_id)
```
Delete this.

ormsbee commented on openedx/core/djangoapps/xblock/api.py:
```diff
@@ -146,65 +155,57 @@ def get_block_metadata(block, includes=()):
+    )
+
+
+def get_library_block_olx(usage_key: LibraryUsageLocatorV2):
```
Move to `content_libraries/api.py`

kdmccormick replied:
@bradenmacdonald given the name, I agree this should being content_libraries/api.py

But nothing about the implementation seems library-specific. Any reason we wouldn't want to just keep it here and convert the signature to:
```python
def get_block_olx(usage_key: UsageKeyV2):
    """
    Get the OLX source of the given XBlock.
    """
```
?

bradenmacdonald replied:
That's even better! Yeah let's leave it here and make it generic like that, if it's not library-specific.

kdmccormick replied:
Done 👍🏻

ormsbee commented on openedx/core/lib/xblock_serializer/api.py:
```diff
@@ -23,5 +23,9 @@ def serialize_modulestore_block_for_blockstore(block):
+
+    TODO: We should bring this up to date with Learning Core. I left the name of
+    this as-is partly because there are bigger questions that I have around how
+    we should alter this (e.g. are we going to remove <xblock-include>?).
```
Add a new implementation and deprecate this one, rather than changing its behavior.

kdmccormick replied:
changed the TODO note to capture this.

ormsbee committed:
refactor: re-order skip and ddt decorators in case that makes any difference

ormsbee commented:
@kdmccormick, @bradenmacdonald: I think I've tracked down at least the package introducing the issue:

This passes on devstack, in the cms image:

```
python -Wd -m pytest -p no:randomly --ds=cms.envs.test openedx/core/djangoapps/content_libraries/
```

This fails:

```
python -Wd -m pytest -p no:randomly --ds=cms.envs.test openedx/core/djangoapps/content_libraries/ openedx/core/djangoapps/course_groups/
```

So something in the `course_groups` tests. I hope that makes finding the problem easier.

I now leave this PR in your care. I don't plan to do any more coding until I get back from vacation. Thank you again!

kdmccormick committed:
refactor: address some minor review comments

* specify 'draft' in docstring for get_library_block
* make create_component_for_block private
* delete leftover get_learning_context_impl comment

kdmccormick committed:
refactor: get_library_block_olx -> get_block_draft_olx

kdmccormick committed:
test: use OpenEdxEventsTestMixin to fix content_libraries event testing

I'm not going to pretend I entirely understand this, but I'd like to get
tests passing first, and then worry about the details.

kdmccormick commented:
Pushed a tentative fix to https://github.com/openedx/edx-platform/pull/34066#issuecomment-1947521522: https://github.com/openedx/edx-platform/pull/34066/commits/246c821545453ffe9d2fa044b5ed7232637dcc0e

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -781,149 +741,143 @@ def create_library_block(library_key, block_type, definition_id):
 
 
-def delete_library_block(usage_key, remove_from_parent=True):
+def component_already_exists(usage_key: UsageKeyV2) -> bool:
```
If this is a public API, it's weird to call it "already" exists, as it may be used in other cases besides creating a component.

This is also another good candidate for moving from the libraries API to the XBlock API. No library-specific code at the moment.

```suggestion
def component_exists(usage_key: UsageKeyV2) -> bool:
```

kdmccormick replied:
Rename is done

kdmccormick replied:
Rather than adding this to the xblock api, I'm going to make it private, so that we can move it later without breaking the public API contract.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -781,149 +741,143 @@ def create_library_block(library_key, block_type, definition_id):
-    Returns a list of LibraryXBlockStaticFile objects.
+    Returns a list of LibraryXBlockStaticFile objects, sorted by path.
+
+    TODO: This is not yet implemented for Learning Core backed libraries.
```
Just noting that when we do implement this, maybe it can be a generic API and not library-specific.

kdmccormick replied:
Added a TODO note to this effect.

bradenmacdonald commented on openedx/core/djangoapps/xblock/api.py:
```diff
@@ -123,7 +132,7 @@ def get_block_metadata(block, includes=()):
         "block_id": str(block.scope_ids.usage_id),
         "block_type": block.scope_ids.block_type,
-        "display_name": get_block_display_name(block),
+        "display_name": block.display_name,
```
This seems like a change - the fallback behavior will be missing if the block hasn't defined a display_name, won't it?

ormsbee replied:
Yeah. I think what happened is that I was originally focusing on the `block_or_key` aspect of it, got rid of the ability for it to be a key, and somewhere through the refactorings dropped the fallback behavior in `get_block_display_name`. But yeah, it should fallback, and in the current code it's not.

kdmccormick replied:
I put the fallback back.

bradenmacdonald commented on openedx/core/djangoapps/xblock/rest_api/views.py:
```diff
@@ -196,7 +200,7 @@ def get(self, request, usage_key_str):
         block = load_block(usage_key, request.user)
         block_dict = {
-            "display_name": get_block_display_name(block),  # potentially duplicated from metadata
+            "display_name": block.display_name,  # potentially duplicated from metadata
```
Same question here - why are we removing the use of `get_block_display_name` ?

bradenmacdonald commented on openedx/core/djangoapps/xblock/runtime/learning_core_runtime.py:
```diff
@@ -0,0 +1,299 @@
+        # We can do this more efficiently in a single query later, but for now
+        # just get it the easy way.
+        component = self._get_component_from_usage_key(usage_key)
+        component_version = component.versioning.draft
```
This runtime needs to support both draft and published versions, and shouldn't be forcing draft version only. (Which one is used should be a property of the runtime when it gets initialized, depending on whether it's running in Studio, LMS, or LMS-Preview.)

We can defer that to a following PR if you want, but if so it needs to be clearly flagged with a TODO.

ormsbee replied:
I had figured it would be puntable for a little while because libraries is Studio only, and we're copying to Modulestore before serving to the LMS.

kdmccormick replied:
Added a TODO comment

bradenmacdonald replied:
We did have two use cases for this already, but I think they're both gone now:
1. LabXchange used the "allow_public_learning" feature, which allows users to view library content via the LMS directly, if they know the URL of an individual component. Basically like a "chromeless" view for use in an iframe, but served by the LMS from the published version, and the LMS can save the learner state.
2. The Library Authoring frontend had an "Preview" tab alongside the "OLX" tab and other tabs, which showed the published LMS version in this same way.

1 is no longer used and 2 is removed I think, so it's fine to punt this for now.

bradenmacdonald reviewed (approved):
Very happy with how this is shaping up :)

Just some minor comments.

ormsbee commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -781,149 +741,143 @@ def create_library_block(library_key, block_type, definition_id):
+        content.id,
+        key="block.xml",
+        learner_downloadable=False
+    )
```
@kdmccormick: My comment about whether this should be in an atomic block is in reference to this part, where we're making a bunch of separate objects for components, component versions, and content here, and this should probably be wrapped together in an `with transaction.atomic():`

kdmccormick replied:
Wrapped

kdmccormick committed:
fix: raise a ValueError if block is set to OLX with the wrong XML tag

kdmccormick committed:
fix: wrap _create_component_for_block in a transaction

kdmccormick committed:
test: use non-ascii characters in library org, slug, and block id

kdmccormick committed:
refactor: don't import MediaType from models

kdmccormick committed:
fix: fall back to type-defined block name if display_name missing

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6214333304.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6214333304-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6214333304-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6214333304-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6215265619.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6215265619-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6215265619-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6215265619-tutor-requirements.txt)

kdmccormick committed:
fix: typo in get_block_display_name

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6215785001.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6215785001-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6215785001-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6215785001-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6216472910.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6216472910-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6216472910-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6216472910-tutor-requirements.txt)

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/tests/test_content_libraries.py:
```diff
@@ -289,12 +240,12 @@ def test_library_blocks(self):
         block_data = self._add_block_to_library(lib_id, "problem", "problem1")
         self.assertDictContainsEntries(block_data, {
-            "id": "lb:CL-TEST:testlib1:problem:problem1",
+            "id": "lb:CL-TعST:testlꜟط:problem:ࠒröblæm1",  # Non-ascii slugs to test unicode robustness.
```
I feel like I'm missing something here... this is checking that the _output_ contains these unicode characters, but the _input_ just a few lines above is ASCII - `problem1` on line 246 for the block_id and `testlib1` on the first line of the function for the library ID. Where are the unicode versions originally coming from?

And is `assertDictContainsEntries` even working correctly?

kdmccormick replied:
@bradenmacdonald the only thing you're missing is that I made the wrong change :) Locally, this test is failing the way you'd expect it to--it hasn't shown up in CI yet because another test failure is cancelling its shard.

kdmccormick committed:
test: fix non-ascii content_libraries tests

kdmccormick commented on openedx/core/djangoapps/content_libraries/serializers.py:
```diff
@@ -128,40 +135,24 @@ class LibraryXBlockTypeSerializer(serializers.Serializer):
+    #
+    # It doesn't look like the frontend actually uses this to put meaningful
+    # slugs at the moment, but hopefully we can change this soon.
     definition_id = serializers.SlugField()
```
1. The vacation energy in this comment's example keys 😆 
2. @bradenmacdonald , tests are currently passing for non-ascii *library slugs*, but are failing for non-ascii *block ids*, because this is a plain old SlugField. To support non-ascii block ids, we'd want:
   ```suggestion
       definition_id = serializers.CharField(validators=(validate_unicode_slug, ))
   ```
   should I go ahead and make this change?

kdmccormick replied:
@bradenmacdonald I've applied this -- let me know if you think I should revert it.

bradenmacdonald replied:
Makes sense to me. Thanks! If there are many other areas that need fixing up, feel free to punt on the unicode slug support though.

kdmccormick committed:
refactor: component_already_exists -> _component_exists

kdmccormick committed:
docs: add TODO comments - draft vs published, and deserialization

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into learning-core-libs-readwrite-runtime

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6225554328.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6225554328-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6225554328-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6225554328-tutor-requirements.txt)

kdmccormick committed:
feat: allow non-ascii block ids

kdmccormick committed:
feat!: rm get_library_blocks

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into learning-core-libs-readwrite-runtime

kdmccormick committed:
fix: typo -- left something commented out while testing

kdmccormick commented:
@bradenmacdonald All comments are addressed. Here are all the changes I've made (plus some noise from the merge commits): https://github.com/ormsbee/edx-platform/compare/0f36be29274ac0fd331eea3954457d8aeaaa5f96...learning-core-libs-readwrite-runtime . Can you give those a look? In the meantime, I'll do some smoke testing.

kdmccormick commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -647,116 +602,109 @@ def _get_library_component_tags_count(library_key) -> dict:
-    # Verify that the OLX parses, at least as generic XML:
+    _block_metadata = get_library_block(usage_key)
+
+    # Verify that the OLX parses, at least as generic XML, and the root tag is correctd:
```
```suggestion
    # Verify that the OLX parses, at least as generic XML, and the root tag is correct:
```

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6226526981.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6226526981-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6226526981-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6226526981-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6226864810.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6226864810-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6226864810-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6226864810-tutor-requirements.txt)

kdmccormick committed:
fix: `only` was conflicting with `select_related`

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6227512249.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6227512249-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6227512249-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6227512249-tutor-requirements.txt)

bradenmacdonald commented on openedx/core/djangoapps/content_tagging/utils.py:
```diff
@@ -93,7 +93,7 @@ def get_library_orgs(self, user, org_names: list[str]) -> list[Organization]:
             library_orgs = {
                 library.org.short_name: library.org
-                for library in get_libraries_for_user(user).select_related('org').only('org')
+                for library in get_libraries_for_user(user).select_related('org').only('org', 'learning_package')
```
Nit: I would appreciate a comment here as a reminder, like:

`# We don't actually need learning_package here but it's already included by get_libraries_for_user() so we need to include it in .only() or we get an ORM error.`

kdmccormick replied:
Yeah good call 👍🏻

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -647,116 +602,109 @@ def _get_library_component_tags_count(library_key) -> dict:
-    BundleCache(metadata.def_key.bundle_uuid, draft_name=DRAFT_NAME).clear()
+    if node.tag != usage_key.block_type:
+        raise ValueError(
+            f"Tried to set the OLX of a {usage_key.block_type} block to a <node.tag> node. "
```
I don't think we want to print a literal `<node.tag>` here? You had `<{node.tag}>` before 12f138b5fe810cd622090ba197cdc3bc499292d6, which makes more sense to me.

kdmccormick replied:
Whoops, yup, ty

bradenmacdonald reviewed (approved):
Those changes look great @kdmccormick 👏🏻

kdmccormick committed:
fix: final review comments

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6228303308.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6228303308-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6228303308-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6228303308-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6233974274.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6233974274-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6233974274-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6233974274-tutor-requirements.txt)

kdmccormick commented:
OK, I haven't the faintest idea why this PR's sandbox is failing migrations, but I did confirm that:

* migrations applied fine on `tutor local`
* migrations applied fine on [this PR's sandbox](https://github.com/openedx/edx-platform/pull/33936), which was originally created using master, and then was updated with this PR's code.
  *  The Blockstore-authored libraries render as title-less cards which spin when you click on them. I think that's completely reasonable for now. If, in the future, we want to degrade more gracefully when there's no backing learning_package for a content_library, then we could do that.
  ![image](https://github.com/openedx/edx-platform/assets/3628148/84ab8375-93df-425b-84af-4a713f758e38)

kdmccormick committed:
Merge branch 'master' into learning-core-libs-readwrite-runtime

kdmccormick committed:
Merge branch 'master' into learning-core-libs-readwrite-runtime

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6235484906.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6235484906-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6235484906-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-34066-2432b1-51564656-6235484906-tutor-requirements.txt)

ormsbee commented:
🎉  @kdmccormick, @bradenmacdonald: Thank you for getting this over the line! It feels so nice to finally have the learning core foundations in place under libraries!

kdmccormick commented on openedx/core/djangoapps/xblock/runtime/learning_core_runtime.py:
```diff
@@ -0,0 +1,302 @@
+        )
+        xml_node = etree.fromstring(content.text)
+        block_type = usage_key.block_type
+        keys = ScopeIds(self.user_id, block_type, None, usage_key)
```
Hey @ormsbee , is setting `def_id=None` here the long-term intention?

Context: I'm on a quest to type-annotate XBlock, and I'm looking for the correct annotation for `def_id`. Currently I have:

```python
DefinitionId: t.TypeAlias = DefinitionKey | UsageKey | ObjectId | LocalId | str
```

and that handles that ModulestoreRuntimes. But in order to handle the LearningCoreRuntime, looks like I'll need to add `| None` to that. Happy to do that, just want to check your instincts before proceeding.

cc @bradenmacdonald

ormsbee replied:
Yeah, it's a long term intention. I think that the `def_id` as it exists today is an unhelpful leak of Split's internal storage details, and that it's more trouble than it's worth. I'd like to eventually kill it off entirely.
